src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs

```diff
@@ -890,5 +891,655 @@ static void Main(string[] args)
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(2).WithArguments("Foo"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(3).WithArguments("Bar"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(4).WithArguments(".ctor"));
+            await test.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestCustomAttribute()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            A aObject = {|#0:new A()|};
+        }
+    }
+
+[My]
+class {|#1:A|}
+{
+}
+
+[RequiresPreviewFeatures]
+[AttributeUsage(AttributeTargets.All)]
+class MyAttribute : Attribute
+{
+}
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments(".ctor"));
+            await test.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestNestedInvocation()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Console.WriteLine({|#0:A.B()|});
+        }
+    }
+
+class A
+{
+    [RequiresPreviewFeatures]
+    public static bool B() => true;
+}
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments("B"));
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestDebug()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+        }
+    }
+
+    public abstract class {|#1:BAbstract|}
+    {
+    }
+}";
+
+            var test = TestCS(csInput);
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestClassImplementingAbstractClassThatImplementsAnInterface()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+    @"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            BImplementation b = new BImplementation();
+            ((IZoo)b).Bar();
+        }
+    }
+
+    public class BImplementation : BAbstract
+    {
+
+    }
+
+    public abstract class BAbstract : IZoo
```

Comment:
Why not warn for every inheritor?

Reply:
I think the rationale was that `IZoo` is not marked as Preview on purpose, so we assume that types deriving/implementing `IZoo` are not Preview. The design calls for 1 level look ups, so it's up to the author to get the attributes applied correctly at every level.

Reply:
sounds good